### PR TITLE
Integrate assignment config

### PR DESCRIPTION
**Pull request by rod-lin**
Reimplemented an old PR https://github.com/illinois-cs241/broadway-on-demand/pull/14
It resolves https://github.com/illinois-cs241/broadway-on-demand/issues/10

@ayushr2

### CODE REVIEW

**Comment by ayushr2** [comment]
The Broadway side of this looks good to me. It is imperative that the assignment ID entered in the textbox is exactly the same as what is passed. The exact same assignment ID should be used which requesting AG runs.

This will prevent issues of uploading configs with wrong URLs and make the entire pipeline of adding an assignment more intuitive.

Note that you should not be able to add an assignment without the config JSON.

**Comment by rod-lin** [comment]
> Note that you should not be able to add an assignment without the config JSON.
Yes that should be the current way of dealing errors from the broadway side.

emm just realized I forgot to add support for assignment edit.

**Comment by ayushr2** [comment]
So now when we edit an assignment, is it optional to edit the config too? Or is it mandatory to upload a new config each time we edit?

**Comment by rod-lin** [comment]
Ohh currently it's gonna update the config every time, even when the config is not edited.
(but I guess I can track changes to the config in the front-end and skip the update if it's not changed at all)

**Comment by ayushr2** [comment]
That is fine. But will it ask the user to upload a config each time? That is my main concern. The user should be given an option to upload a new config for the assignment. If they do then use the new config, if not use the old one. If changes are only made to the start/end date of assignment, request to Broadway need not be made.

**Comment by rod-lin** [comment]
Oh actually I'm making a request to broadway to fetch the current config every time the assignment page is loaded. So the page will fill in the old config automatically.

**Comment by rod-lin** [comment]
I added a new endpoint `GET /staff/config/<cid>/<aid>/` to dynamically load old configuration

File: src/templates/staff/assignment.html [hunk 1]
```diff
@@ -115,6 +132,30 @@
...
+        $.ajax({
+            type: "GET",
+            url: "{{ url_for('.staff_get_assignment_config', cid=course._id, aid=assignment.assignment_id) }}",
+            dataType: "json",
+            success: function (result) {
+                let json = JSON.stringify(result);
+                let old = normalizeJSON(json);
+                $('#mdl-edit-assn-config').val(old).data("old-config", old).removeAttr("disabled");
+            },
+            error: function () {
```
Comment (shreyas208) [inline src/templates/staff/assignment.html]: If loading the old config failed, I think we should keep the config field disabled and show a message like "Config loading failed; you may proceed with editing other information". If something is broken, we should be in a fail-safe state and not allow config changes via on-demand as there is probably a version inconsistency between on-demand and API, or API is down.

Comment (rod-lin) [reply src/templates/staff/assignment.html]: Yeah that makes sense

**Comment by rod-lin** [comment]
Okay fixed.

@shreyas208 I can add it tomorrow night if it's not too urgent

File: src/bw_api.py [hunk 2]
```diff
@@ -71,7 +71,8 @@ def get_assignment_config(cid, aid):
 		if resp.status_code == 200:
 			ret_data = resp.json()
 			return ret_data["data"]
-	except: pass
+	except Exception as e:
```
Comment (shreyas208) [inline src/bw_api.py]: I'm guessing you looked at an old commit, the current code has something like this:
`logging.error("start_grading_run(cid={}, aid={}, netid={}): {}".format(cid, aid, netid, repr(e)))
`

**Comment by shreyas208** [comment]
Tomorrow is fine. Thanks!

File: src/bw_api.py (hunk 2 above)
Comment (rod-lin) [reply src/bw_api.py]: ohhh yeah

**Comment by rod-lin** [comment]
@shreyas208 I added the file upload. But I'm only reading the file on the client side instead of uploading it to the server with form data.

**Comment by shreyas208** [comment]
@rod-lin that's the method I was thinking of too :+1: reviewing

**Comment by shreyas208** [comment]
Looks good!

**Pull request merged**